File: recursion_i/recurrence_relation/pascals_triangle_ii/pascals_triangle_ii.cpp
/**
 * Author: Nathan Bockisch
 * Date: October 21, 2021
 *
 * Explanation:
 * Pascal's triangle is an array of binomial coefficients,
 * a value at row a and col b is a! / b!(a - b)!. Binomial
 * coefficients have an additive property, so the value
 * (a, b) = (a - 1 / b - 1) + (a - 1 / b). SUccessive binomial 
 * coefficients differ by (a! / b!(a - b)!) / (a! / (b - 1!)(a - b + 1)!) 
 * = (a = b + 1) / b
 *
 * Therefore each element in the row is the previous element multiplied by 
 * this coefficient.
 **/

class Solution {
public:
    vector<int> getRow(int rowIndex) {
        vector<int> row = {1};
        
        for (int i = 1; i <= rowIndex; i++) 
            row.push_back((int)((row.back() * (long long)(rowIndex - i + 1)) / i));        
        
        return row;
    }
};

File: array_and_string/conclusion/pascals_triangle_ii/pascals_triangle_ii.rs
/*
 * Author: Nathan Bockisch
 * Date: July 15, 2022
 */
impl Solution {
    pub fn get_row(row_index: i32) -> Vec<i32> {
        let row_index = row_index as usize;
        let mut answer = vec![1; row_index + 1];

        for i in 1..row_index {
            let mut j = i;
            while (j > 0) {
                answer[j] += answer[j - 1];
                j -= 1;
            }
        }

        answer
    }
}
